### ISSUE TITLE
Race condition while shutting down process-compose

### ISSUE THREAD

**Issue by zmitchell** [F1bonacc1/process-compose#197]
## Defect

Make sure that these boxes are checked before submitting your issue -- thank you!

- [x] Included the relevant configuration snippet
- [x] Included the relevant process-compose log (log location: `process-compose info`)
- [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Version of `process-compose`:
Verified on main branch and v1.6.1

#### OS environment:
macOS, Linux

#### Steps or code to reproduce the issue:
Every now and then (seemingly at random) I would see this error when calling `process-compose down`:
```
24-07-11 14:59:45.839 FTL failed to stop project error="Post \"http://unix/project/stop\": EOF"
```
I thought it was probably me misconfiguring something, but I've come to realize it's a race condition in this function in `pc_api.go`:
```go
func (api *PcApi) ShutDownProject(c *gin.Context) {
	api.project.ShutDownProject()
	c.JSON(http.StatusOK, gin.H{"status": "stopped"})
}
```
I suspect there is a race between shutting down the processes and actually sending the response.

Since this is a race condition I suspect you run into it more often when you have processes that are very quick to shut down. This is the case in the scenarios I've encountered the issue:
```yaml
# service-config.yaml
processes:
  sleeping1:
    command: "sleep 999999"
  sleeping2:
    command: "sleep 999999"
```

In order to reproduce this error you can run `process-compose` under a debugger in one terminal and make calls from the client in another terminal.
```text
# term1
$ dlv debug github.com/f1bonacc1/process-compose/src -- -f service-config.yaml -u pc.sock --tui=false
...
> break pc_api.go:270
> c
```
The debugger will now pause while `process-compose` waits for API calls.
```text
# term2
$ process-compose down -u pc.sock
```
`term2` will now hang (I suppose it's waiting for a response) and `term1` will proceed until it hits the breakpoint. If you go back to `term1` and kill `process-compose` by leaving the debugger, you'll see the exact error I mentioned above.
```text
# term1
> quit
```
```text
# term2
24-07-11 14:59:45.839 FTL failed to stop project error="Post \"http://unix/project/stop\": EOF"
```

#### Expected result:
`process-compose` should reliably and gracefully shutdown when requested.

#### Actual result:
In some cases, `process-compose` does not shut down cleanly.

**Comment by F1bonacc1** [comment]
Fixed in v1.9.0

**Comment by zmitchell** [comment]
I'm not sure this actually fixed the problem. It appears that there's _also_ a race condition on start:
- `process-compose up -f my-config.yaml -u /insufficient_perms.sock`
- which one wins?
    - Starting processes
    - Attempting to connect to that socket

@dcarley and I were seeing that issue before #201, and not after, but @dcarley also mentioned that we may still be seeing the original shutdown race condition. We'll wait to confirm until @dcarley has a reproducer.
